spanconfigsqltranslator: populate protected_timestamps in SpanConfig

adityamaru opened the pull request:
This change teaches the SQLTranslator to hydrate the SpanConfigs
for a table with protected timestamps that apply to that table.

Concretely, this change initializes a spanconfig.ProtectedTimestampStateReader
in the txn in which the translation is taking place, thereby
providing a transactional view of the system.protected_ts_records
table. After generating the span configurations based on the zone
configurations that apply to the table, we hydrate the newly
introduced protected_timestamps field on each span configuration
with all the protected timestamps that apply to this table. This
includes protected timestamp records that directly target this
table, as well as records targetting the table's parent database.
This information is obtained from the ProtectedTimestampStateReader
mentioned above.

Additionally, this change modifies StartTenant to allow secondary
tenants to interact with the protected timestamp subsystem using a
"real" protectedts.Provider provided the migration
EnableProtectedTimestampsForTenant has run.

For testing purposes, this change teaches the data driven framework
of two additional commands protect and release.

Informs: #73727

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74803)
<!-- Reviewable:end -->

adityamaru commented:
First two commits are from https://github.com/cockroachdb/cockroach/pull/74737.

adityamaru committed:
spanconfigsqltranslator: populate protected_timestamps in SpanConfig

This change teaches the SQLTranslator to hydrate the SpanConfigs
for a table with protected timestamps that apply to that table.

Concretely, this change initializes a spanconfig.ProtectedTimestampStateReader
in the txn in which the translation is taking place, thereby
providing a transactional view of the system.protected_ts_records
table. After generating the span configurations based on the zone
configurations that apply to the table, we hydrate the newly
introduced protected_timestamps field on each span configuration
with all the protected timestamps that apply to this table. This
includes protected timestamp records that directly target this
table, as well as records targetting the table's parent database.
This information is obtained from the ProtectedTimestampStateReader
mentioned above.

Additionally, this change modifies StartTenant to allow secondary
tenants to interact with the protected timestamp subsystem using a
"real" protectedts.Provider provided the migration
EnableProtectedTimestampsForTenant has run.

For testing purposes, this change teaches the data driven framework
of two additional commands protect and release.

Informs: #73727

Release note: None

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -140,6 +141,11 @@ message SpanConfig {
+  // ProtectedTimestamps captures all the protected timestamp records that apply
+  // to the range. The timestamp values represent the timestamps after which
+  // data will be protected from GC, if the protection succeeds.
+  repeated util.hlc.Timestamp protected_timestamps = 10 [(gogoproto.nullable) = false];
```
this is subject to change in https://github.com/cockroachdb/cockroach/pull/74765

irfansharif commented on pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go:
```diff
@@ -279,6 +338,7 @@ func (s *SQLTranslator) generateSpanConfigurationsForTable(
 			})
 		}
 
+		hydrateSpanConfigurationsForTable(&entries, desc, ptsStateReader)
```
Instead of this, could we perhaps just populate in tableSpanConfig right when it's constructed? Would the following suffice:

```go
	tableSpanConfig := zone.AsSpanConfig()
	tableSpanConfig.ProtectedTimestamps = append(tableSpanConfig.ProtectedTimestamps,
		ptsStateReader.GetProtectedTimestampsForSchemaObject(desc.GetID())...)
	tableSpanConfig.ProtectedTimestamps = append(tableSpanConfig.ProtectedTimestamps,
		ptsStateReader.GetProtectedTimestampsForSchemaObject(desc.GetParentID())...)
```

Below, when grafting the subzone config, just be sure to copy over the array verbatim (we can't protect a subzone, right?).

irfansharif replied:
Let's make this more involved than it needs to be in the commit that actually requires it. This method at large can probably be broken down, and it might happen with the introduction of a spanconfig.Splitter interface I'm trying to extract for https://github.com/cockroachdb/cockroach/issues/70555; we can revisit once we're ready to land #73925 .

adityamaru replied:
Done as discussed offline, looks much cleaner!

irfansharif commented on pkg/spanconfig/spanconfigtestutils/utils.go:
```diff
@@ -236,6 +240,19 @@ func PrintSpanConfigDiffedAgainstDefaults(conf roachpb.SpanConfig) string {
 		diffs = append(diffs, fmt.Sprintf("lease_preferences=%v", conf.VoterConstraints))
 	}
+	if !reflect.DeepEqual(conf.ProtectedTimestamps, defaultConf.ProtectedTimestamps) {
+		sort.Slice(conf.ProtectedTimestamps, func(i, j int) bool {
```
Can we avoid sorting? I think this mutates the underlying array, and I'm not sure we need it for test determinism.

irfansharif commented on pkg/spanconfig/spanconfigtestutils/utils.go:
```diff
@@ -236,6 +240,19 @@ func PrintSpanConfigDiffedAgainstDefaults(conf roachpb.SpanConfig) string {
+		sort.Slice(conf.ProtectedTimestamps, func(i, j int) bool {
+			return conf.ProtectedTimestamps[i].Less(conf.ProtectedTimestamps[j])
+		})
+		var s string
```
Instead of string concatenation in this form, we could collect an array of strings and use something like: `fmt.Sprint("pts=[%s]", strings.Join(timestamps, " ")`

irfansharif commented on pkg/spanconfig/spanconfigtestutils/utils.go:
```diff
@@ -324,3 +341,45 @@ func GetSplitPoints(ctx context.Context, t *testing.T, reader spanconfig.StoreRe
+		}
+		return ptpb.MakeSchemaObjectsTarget(ids)
+	default:
+		t.Fatalf("malformed line %q, expecetd to find prefix %q or %q", target, tenantPrefix,
```
[nit] Typo: "expecetd". Also, what about cluster prefix?

adityamaru commented on pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go:
```diff
@@ -98,6 +120,29 @@ func TestDataDriven(t *testing.T) {
+		mkRecordAndProtect := func(recordID string, ts hlc.Timestamp, target *ptpb.Target) {
+			jobID := jr.MakeJobID()
+			require.NoError(t, execCfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) (err error) {
+				recID, err := uuid.FromBytes([]byte(strings.Repeat(recordID, 16)))
```
nice catch, asserted that the ID is len 1. If and when we find ourself needing more we can tweak this.

adityamaru commented:
@arulajmani @irfansharif this should be RFAL!

irfansharif commented on pkg/server/testserver.go:
```diff
@@ -500,12 +501,18 @@ func (ts *TestServer) Start(ctx context.Context) error {
+	if d.st.Version.IsActive(ctx, clusterversion.EnableProtectedTimestampsForTenant) {
+		return d.Provider.Protect(ctx, txn, rec)
+	}
+	return errors.New("tenant cannot Protect timestamp until the clusterversion has been finalized")
```
Something a bit more informative perhaps? `"%s is inactive", clusterversion.EnableProtectedTimestampsForTenant`. Also [nit] flip the conditional so the error case is handled above above and the happy path here, outdented.

irfansharif commented on pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go:
```diff
@@ -98,6 +117,10 @@ func TestDataDriven(t *testing.T) {
 			tenant = spanConfigTestCluster.InitializeTenant(ctx, roachpb.SystemTenantID)
 		}
 
+		execCfg := tenant.ExecCfg()
```
[nit] Remove these intermediate variables, using the symbols on the Tenant handle below directly is no less clear.

irfansharif reviewed (approved):
LGTM mod comments below.

adityamaru commented on pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go:
```diff
@@ -61,6 +67,15 @@ import (
+// - "protect" [record-id=<int>] [ts=<int>]
+//   cluster                  OR
+//   tenants       id1,id2... OR
+//   descs  id1,id2...
```
whoops, fixed.

adityamaru commented on pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go:
```diff
@@ -77,17 +92,21 @@ func TestDataDriven(t *testing.T) {
 		ManagerDisableJobCreation: true,
 	}
+	ptsKnobs := &protectedts.TestingKnobs{
+		EnableProtectedTimestampForMultiTenant: true,
```
Added a TODO to delete this knob from the datadriven test once it defaults to true. Maybe im misunderstanding but we'd still need to pull in the package import when populating the PTSKnobs here https://github.com/adityamaru/cockroach/blob/full-translate-contd/pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/cluster.go#L73-L73.

arulajmani reviewed (approved):
> Reviewed 2 of 22 files at r7, 19 of 22 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74803)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @irfansharif)

___
*[pkg/clusterversion/cockroach_versions.go, line 386 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74803#-MuGklIwCfa2DR8s31mr:-MuGklIwCfa2DR8s31ms:b17acui) ([raw file](https://github.com/cockroachdb/cockroach/blob/29a4a036670a85caa76ee0f164de923cfe069506/pkg/clusterversion/cockroach_versions.go#L386)):*
> ```Go
> 		Version: roachpb.Version{Major: 21, Minor: 2, Internal: 48},
> 	},
> 	{
> ```

To confirm, it's okay for us to do this now because we haven't released any betas? 
___
*[pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go, line 376 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74803#-MuGpl_D9AAYBv00eVXh:-MuGpl_D9AAYBv00eVXi:b-z54k2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/29a4a036670a85caa76ee0f164de923cfe069506/pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go#L376)):*
> ```Go
> 		// Copy the ProtectionPolicies that apply to the table's SpanConfig onto its
> 		// SubzoneSpanConfig.
> 		subzoneSpanConfig.GCPolicy.ProtectionPolicies = tableSpanConfig.GCPolicy.ProtectionPolicies[:]
> ```

nit: would it be clearer if you pulled the right hand side of the assignment in a separate variable and used it to assign the ProtectionPolicies wherever you need to?


<!-- Sent from Reviewable.io -->

irfansharif commented:
> To confirm, it's okay for us to do this now because we haven't released any betas?

Yes, but also because we don't support upgrades from alphas/betas to prod clusters. Re-ordering these versions is safe any time before the actual release itself (not even the branch cut).

irfansharif commented on pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go:
```diff
@@ -183,6 +203,37 @@ func TestDataDriven(t *testing.T) {
+				require.NoError(t, tenant.ExecCfg().DB.Txn(ctx,
+					func(ctx context.Context, txn *kv.Txn) error {
+						require.Len(t, recordID, 1,
+							"datadriven test only supports single digit record IDs")
```
[nit] Just to make these datadriven tests clearer ("protect record-id=2 ts=2" for e.g.), I think using characters here would be better.

irfansharif commented on pkg/sql/logictest/testdata/logic_test/crdb_internal_tenant:
```diff
@@ -11,6 +11,8 @@ SELECT node_id, name FROM crdb_internal.leases ORDER BY name
 0  eventlog
 0  jobs
 0  locations
+0  protected_ts_meta
```
BTW, what's this about?

adityamaru replied:
I'm not a 100% sure, but a little digging led me to https://github.com/cockroachdb/cockroach/pull/69008 where we switched to leasing all system descriptors except a few in a deny list. The PTS tables are not in this deny list so I guess we expect to see them here?

irfansharif commented on pkg/spanconfig/spanconfigsqltranslator/protectedts_state_reader.go:
```diff
@@ -33,59 +32,60 @@ type protectedTimestampStateReader struct {
 }
 
-// GetProtectedTimestampsForCluster returns all the protected timestamps that
+// GetProtectionPoliciesForCluster returns all the protected timestamps that
```
[nit] None of these methods need to be public.

adityamaru commented:
>nit: would it be clearer if you pulled the right hand side of the assignment in a separate variable and used it to assign the ProtectionPolicies wherever you need to?

Discussed offline, going to keep it as is for now since it avoids an extra copy.

adityamaru commented:
Bazel timeout seems unrelated.
TFTRs!

bors r=irfansharif,arulajmani
